### ISSUE TITLE
Babel 7 support

### ISSUE THREAD

**Issue by verydanny** [parcel-bundler/parcel#868]
## Feature request

Support Babel 7

**Comment by davidnagli** [comment]
Ya we should definitely work towards introducing support for Babel 7.

The weird thing that surprises me is that you didn’t get an error! 

https://github.com/parcel-bundler/parcel/blob/cafb6ef8f96f0159f22de9e3eb89773709ec9e50/src/transforms/babel.js#L44

We currently have a check that throws an error whenever you’re using an unsupported Babel version, so I’m thinking maybe your error is getting caused by something else it’s hitting before even starting the Babel transform. Did you get a stack trace with the error?

**Comment by DeMoorJasper** [comment]
@davidnagli possibly because he didn't have babel 7 installed locally

**Comment by verydanny** [comment]
Stack trace:

```bash
Server running at http://localhost:1234 
🚨  /Users/daniel/Git/21g/1zet/src/index.js: Cannot read property 'loose' of undefined (While     at _default (/Users/daniel/Git/21g/1zet/node_modules/@babel/plugin-transform-modules-commonjs/lib/index.js:21:23)
    at Function.memoisePluginContainer (/Users/daniel/Git/21g/1zet/node_modules/babel-core/lib/transformation/file/options/option-manager.js:113:13)
    at Function.normalisePlugin (/Users/daniel/Git/21g/1zet/node_modules/babel-core/lib/transformation/file/options/option-manager.js:146:32)
    at /Users/daniel/Git/21g/1zet/node_modules/babel-core/lib/transformation/file/options/option-manager.js:184:30
    at Array.map (<anonymous>)
    at Function.normalisePlugins (/Users/daniel/Git/21g/1zet/node_modules/babel-core/lib/transformation/file/options/option-manager.js:158:20)
    at OptionManager.mergeOptions (/Users/daniel/Git/21g/1zet/node_modules/babel-core/lib/transformation/file/options/option-manager.js:234:36)
    at /Users/daniel/Git/21g/1zet/node_modules/babel-core/lib/transformation/file/options/option-manager.js:265:14
    at /Users/daniel/Git/21g/1zet/node_modules/babel-core/lib/transformation/file/options/option-manager.js:323:22
    at Array.map (<anonymous>)
```

**Comment by DeMoorJasper** [comment]
@verydanny do you get the `Unsupported babel version: ${babel.version}` error if u install babel 7 in your project

**Comment by verydanny** [comment]
I do not, and I do have babel 7 installed locally.

**Comment by devongovett** [comment]
Ya, because the package name is `@babel/core` not `babel-core` in babel 7, so the error wouldn't occur.

**Comment by devongovett** [comment]
So, we need to figure out how to support both babel 7 and babel 6 simultaneously I think. Parcel will need to build projects built with Babel 6 for a while now, but we should also support Babel 7 projects and perhaps use Babel 7 internally.

Here's what I think we should do:

1. Once Babel 7 releases a final version, switch to it internally for parsing, running preset-env, transforming JSX, etc.
2. If there is a custom `.babelrc` config, detect the locally installed babel version in the app (babel-core version in package.json). We will need to require that `babel-core` or `@babel/core` is installed locally in the app if you want to use a custom babelrc.
3. If the babel version installed in the app differs from the internal babel version used by Parcel, do two separate compiles: 1 for the internal transforms using babel 7, and one for the custom babelrc. This may require a duplicate parse + codegen step if the Babel 7 ast is incompatible with Babel 6. 😦 

This doesn't sound great actually. I wonder if we can get around it somehow...

**Comment by lifeiscontent** [comment]
@devongovett this might also help with typescript since babel7 has first class support for it now!

**Comment by elken** [comment]
I think a buildtime flag (for now at least) will resolve it.

`parcel --with-babel7` or something similar.

**Comment by DeMoorJasper** [comment]
@elken That won't be neccessary if anyone would like to add babel 7 support we could just detect the version (the same way we already do to warn unsupported babel versions, with adding `@babel/core` to the check obviously) and just try to support both.
Not sure how we'd achieve this, Devon already explained how he would approach it. Would be nice to not have to parse and generate 2 times.

**Comment by wachunei** [comment]
I'm literally just starting to learn parcel 😄. Is this issue the reason I come up with this error? 
![screen shot 2018-04-09 at 11 16 16 am](https://user-images.githubusercontent.com/1024246/38503196-d612f94c-3be7-11e8-985a-315c922f5f13.png)

If so..

Why is `@babel/helper-...` being used by parcel dependency `babel-core` instead of a `babel-helper-...` compatible version? It seems like at this point it is necessary to have a duplication of old/new dependencies for internal/external babel transformation. Is this what Devon means? Sorry for the questions, I'm trying to get it right and help 😅

**Comment by DeMoorJasper** [comment]
It probably is due to parcel not supporting babel 7.
It would be nice to find a way to hotswap between `babel-core` and `@babel/core` without losing much of the performance.

**Comment by pravdomil** [comment]
https://github.com/parcel-bundler/parcel/pull/475

**Comment by babakness** [comment]
Would be nice to have this, perhaps with a flag?

**Comment by vihanb** [comment]
adding babel core as peer-dep will allow users to have `babel-core@7.0.0-bridge.0` meaning they can use `@babel/core` (v7) without dropping support for v6

**Comment by reccanti** [comment]
Any update on this?

**Comment by DeMoorJasper** [comment]
@jeroenptrs we don’t really use those boards a lot. If a contributor wants to dig into this issue a PR will pop up, but my guess is it’ll be part of parcel 2

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by rafaelrinaldi** [comment]
I agree with @devongovett on waiting for a stable release of Babel 7. There's still quite a lot of things changing and trying to keep up with it might not be worth it now.

**Comment by verekia** [comment]
Is there any workaround on this issue at all? Even something very ugly? Or is the only option to use a Babel 7 stack to switch to Webpack in the meantime...?

**Comment by DeMoorJasper** [comment]
@verekia the workaround is using babel 6, babel 7 is still in Beta that’s why we dont support it yet...

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by verekia** [comment]
How about just supporting the bridge? That's exactly what it's for, to help libraries with the transition.

https://github.com/babel/babel-bridge

**Comment by DeMoorJasper** [comment]
Well in this case `babel-bridge` is fairly useless as that just creates a wrapper to not break package.json. The main issue is that it doesn't account for any breaking changes in the API as far as I know and parcel relies very heavily on Babel, and that's the actual challenging part about it.

**Comment by verekia** [comment]
Okay thanks for your insight :) I guess we'll wait!

**Comment by adrienharnay** [comment]
On the bright side ☀️, Babel 7 is around the corner!

**Comment by olstenlarck** [comment]
Hey @DeMoorJasper what's needed for that to land?

Babel 7 is mostly here and all of the scoped packages are as `latest`.

**Comment by DeMoorJasper** [comment]
@olstenlarck the main issue is we have to make parcel support both babel 6 and 7 without making it much slower or complex.

Parcel's internals should probably remain to use babel 6 untill it's out of beta, would be nice if we would be able to switch instantly once babel 7 lands though.

You could look through this thread to see some discussion around implementation and opinions from the parcel team. It should give you a nice overview. Feel free to ask any more questions :)

**Comment by adrienharnay** [comment]
Hey @DeMoorJasper, to make it sure I understood your answer: once Babel 7 is out of beta, Parcel will switch to it (most likely with a major release)?

Thanks!

**Comment by DeMoorJasper** [comment]
First RC for Babel 7 has been released a couple hours ago as mentioned in #475 :)

If anyone is interested in researching how to support both 6 and 7 feel free to do so

**Comment by vihanb** [comment]
@DeMoorJasper I'm not too familiar with Parcel internals but couldn't babel7 be invoked based on presence of `@babel/core` module?

**Comment by adrienharnay** [comment]
I'm not sure this can be done/is worth to fight for. The majority of projects I know have released a major version, and stopped supporting Babel6 since. However, if you think this could be achieved, I would be happy to lend a hand!

**Comment by DeMoorJasper** [comment]
@adrienharnay it would be cool if we could support backwards compatibility otherwise it might be part of parcel 2 which isn’t gonna get released any time soon

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by olstenlarck** [comment]
>  and otherwise runs legacy code?

No.

> is worth to fight for

Exactly, not worth.

If backward compat is so important... allow it through flag (`--babel6`?)for some months.

**Comment by DeMoorJasper** [comment]
@olstenlarck The flag seems to be a bit obsolete. The biggest challenge is making parcel support both, not detecting which one to use (we can probably detect it at the root and pass it along everywhere in the options object as described below).

If the user defines a babelrc file and/or has babel installed inside pkg.json you can derive pretty easily whether it's babel 6 or 7

**Comment by olstenlarck** [comment]
> First RC for Babel 7 has been released a couple hours ago as mentioned in #475 :)

@DeMoorJasper sooo, any way to try the things? It was closed so...

**Comment by DeMoorJasper** [comment]
@olstenlarck the PR wasn't working perfectly and was no longer up to date.

We'll send an update here whenever we have anything worthy of testing

**Comment by adrienharnay** [comment]
Babel 7 has been released 😬

**Comment by DeMoorJasper** [comment]
Awesome, Babel 7 is finally out :)

Babel 7 support will be part of Parcel 2 as it will have the ability to check AST compatibility and stringify and parse based on whether it's compatible.

Not sure if it's possible to hack Babel 7 support into Parcel 1, but if anyone is up for the challenge feel free to try. (You would also have to stringify and parse and check compatibility, but I doubt you'd be able to make it possible in a clean way).

**Comment by devongovett** [comment]
WIP here: https://github.com/parcel-bundler/parcel/pull/1955. If you want to help, check out the to do list there and make PRs against the babel7 branch. Thanks!

**Comment by scrossan-crown** [comment]
I was wondering since babel 7 adds support for TypeScript would it possible to add type checking to parcel under a flag. Also it would use the specified version of TypeScript in the projects package.json. The reason I ask for this is I've been using https://github.com/fathyb/parcel-plugin-typescript for a while now but there hasn't been any activity on the project for around a couple of months and it would be great if it was built in.

**Comment by wongmjane** [comment]
Since Babel 7 supports transpiling TypeScript files, it could make transpiling in itself easier - let the Babel 7 compiler handle it, while tsc will only be used for type checking

Related Articles:

https://babeljs.io/blog/2018/08/27/7.0.0#typescript-support-babel-preset-typescript

https://blogs.msdn.microsoft.com/typescript/2018/08/27/typescript-and-babel-7/

**Comment by DeMoorJasper** [comment]
Using babel for ts would probably speed up compiling and fix some bugs in the process. However I think we should start by focusing on making babel 7 work properly for js

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by HauntedSmores** [comment]
Ran into this issue when trying to use Parcel's internal version of babel for my js but a different preset to run a gulpfile with ES6. 

Is there a way to use a different `.babelrc` for Parcel vs `gulpfile.babel.js`?

**Comment by rafaelrinaldi** [comment]
@HauntedSmores For reference I know Rollup has a `babelrc` flag which you can opt out of. Might be worth checking if something like that makes sense in Parcel as well (I think it does).

**Comment by ccampean** [comment]
As devDependencies I have installed 
```json
    "@babel/core": "^7.0.0",
    "babel-core": "^7.0.0-bridge"
    "parcel-bundler": "^1.9.7",
```
As peerDependencies I have this 
```json
"babel-core": "6.x || ^7.0.0-bridge.0"
```
But I still get the error __Requires Babel "^7.0.0-0", but was loaded with "6.26.3"__ How can I fix it?

**Comment by olstenlarck** [comment]
You can't, it's internal thing and we should wait... ;/

**Comment by JacobMGEvans** [comment]
Following.

**Comment by codeaholicguy** [comment]
How is the progress of this? Babel 7.0 released 14 days ago. 😭

**Comment by shofel** [comment]
@codeaholicguy Nguyễn, You're welcome helping out with this! :) As I believe, #1955 is the starting point to participate on babel 7 support.

**Comment by devongovett** [comment]
Merged #1955 and released a beta in v1.10.0-beta.0. Please help test it out and report any bugs you find! 🎉

**Comment by devongovett** [comment]
I should have given some info about how it works.

* For internally generated configs, e.g. our built in babel-preset-env, and JSX support, we will use babel 7 automatically. If you don't have a `.babelrc` with custom plugins in it, you don't need to do anything at all.
* On existing projects with a `.babelrc` using babel 6, we will continue using babel 6 as normal. If you don't have an explicit dependency on `babel-core` in your package.json, one will be automatically installed for you.
* If you would like to move a project from babel 6 to 7, you need to add a dependency on `@babel/core`, and adjust your `.babelrc` to point to babel 7 plugins.
* For new projects, when you create a `.babelrc` file, we will automatically detect whether the plugins are dependent on babel 6 or 7 and automatically install either `babel-core` or `@babel/core` into your project for you. We also now support automatically installing the plugins themselves automatically as well. 💥

**Comment by qm3ster** [comment]
@devongovett 
> automatically installing the plugins themselves automatically

**Comment by Mike1stver** [comment]
I my friends, I have just create a project in babel by using this commands:

npm install -g parcel-bundler and
npm init -y

and npm install react react-dom.

As i understand, parcel will automatically use babel 7 for my configuration right? I did not use the .babelrc at all, however, i cannot use class properties i have this error:

 Support for the experimental syntax 'classProperties' isn't currently enabled (5:19):

As i understand if parcel , in the way i configured, established babel 7 and not 6 for my project, @babel/preset-env will be installed and i do not need anything else to make the callProperties to works okey. How can I correctly select the version of babel? it seems the v6 is instaled and not the 7. which is the version a need indeed

**Comment by tunnckoCore** [comment]
There are two things: a) using v7 may still not be the default, or b) i'm not sure if the preset-env includes the class properties.
